File: SampleDirectXForms/D3DApp.cs
namespace SampleDirectXForms;

using static NativeMethods;

internal static class D3DApp
{
    public static void Run()
    {
        while (GetMessage(out var msg, nint.Zero, 0, 0) != 0)
        {
            TranslateMessage(ref msg);
            DispatchMessage(ref msg);
        }
    }

    public static int Run(D3DGame game)
    {
        game.Init();

        var msg = new NativeMessage();

        while (msg.msg != WM_QUIT)
        {
            if (PeekMessage(out msg, nint.Zero, 0, 0, (uint)PeekMessageParams.PM_REMOVE))
            {
                TranslateMessage(ref msg);
                DispatchMessage(ref msg);
            }
            else
            {
                game.Tick();
            }
        }

        return msg.wParam.ToInt32();
    }
}

File: SampleDirectXNative/D3DApp.cpp
#include "D3DApp.h"

int D3DApp::Run(D3DGame game)
{
    game.Init();

    MSG msg = { nullptr };

    while (msg.message != WM_QUIT)
    {
        if (PeekMessage(&msg, nullptr, 0, 0, PM_REMOVE))
        {
            TranslateMessage(&msg);
            DispatchMessage(&msg);
        }
        else
        {
            game.Tick();
        }
    }

    return static_cast<int>(msg.wParam);
}
